### PR TITLE
GLideNUI: detect application instance dynamically

### PR DESCRIPTION
**Pull request by Rosalie241**
tested on windows & linux on both project64 and mupen64plus

### CODE REVIEW

File: src/GLideNUI/GLideNUI.cpp [hunk 1]
```diff
@@ -29,27 +29,32 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
 	if (config.generalEmulation.enableCustomSettings != 0 && _romName != nullptr && strlen(_romName) != 0)
 		loadCustomRomSettings(strIniFileName, _romName);
 
-#ifndef M64P_GLIDENUI
-	int argc = 0;
-	char * argv = 0;
-	QApplication a(argc, &argv);
+	QCoreApplication* app = QCoreApplication::instance();
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:32]: QCoreApplication* pApp

File: src/GLideNUI/GLideNUI.cpp [hunk 2]
```diff
@@ -29,27 +29,32 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
 
-	QTranslator translator;
-	if (translator.load(getTranslationFile(), strIniFileName))
-		a.installTranslator(&translator);
-#endif // M64P_GLIDENUI
+	if (app == nullptr)
+	{
+		int argc = 0;
+		char * argv = 0;
+		app = new QApplication(argc, &argv);
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:38]: When and how this object destroyed?
I don't see delete app.

File: src/GLideNUI/GLideNUI.cpp [hunk 3]
```diff
@@ -29,27 +29,32 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
-#ifndef M64P_GLIDENUI
 	w.show();
-	const int res = a.exec();
-#else
-	const int res = w.exec();
-#endif // M64P_GLIDENUI
+
+	int res;
+	if (app == nullptr)
+		res = app->exec();
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:54]: ???!!!

File: src/GLideNUI/GLideNUI.cpp [hunk 4]
```diff
@@ -29,27 +29,32 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
-	int argc = 0;
-	char * argv = 0;
-	QApplication a(argc, &argv);
+	QCoreApplication* app = QCoreApplication::instance();
 
-	QTranslator translator;
-	if (translator.load(getTranslationFile(), strIniFileName))
-		a.installTranslator(&translator);
-#endif // M64P_GLIDENUI
+	if (app == nullptr)
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:34]: if (app == nullptr) {  // code formatting style

File: src/GLideNUI/GLideNUI.cpp [hunk 5]
```diff
@@ -29,27 +29,32 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
+	{
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:35]: if (app == nullptr) { // code formatting style

File: src/GLideNUI/GLideNUI.cpp (hunk 3 above)
Comment (Rosalie241) [reply src/GLideNUI/GLideNUI.cpp:54]: oh, right, this is wrong, will fix!

**Comment by Rosalie241** [comment]
I'll need to test this (on project64), will do when the appveyor build is finished

File: src/GLideNUI/GLideNUI.cpp [hunk 6]
```diff
@@ -29,27 +29,34 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
-#ifndef M64P_GLIDENUI
 	w.show();
-	const int res = a.exec();
-#else
-	const int res = w.exec();
-#endif // M64P_GLIDENUI
+
+	int res;
+	if (!hasApp) {
+		res = pApp->exec();
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:54]: again? !hasApp means pApp is null, so pApp->exec() will creash.

Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:54]: Do I get it correct that you are using QCoreApplication::instance() to find if plugin is in zilmar-spec mode or in mupen64plus one? If QCoreApplication::instance() returns valid pointer, the plugin is in M64P_GLIDENUI mode and thus w.exec() should be called?

File: src/GLideNUI/GLideNUI.cpp [hunk 7]
```diff
@@ -29,27 +29,34 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
-		a.installTranslator(&translator);
-#endif // M64P_GLIDENUI
+	if (!hasApp) {
+		int argc = 0;
+		char * argv = 0;
+		pApp = new QApplication(argc, &argv);
+
+		QTranslator translator;
+		if (translator.load(getTranslationFile(), strIniFileName))
+			pApp->installTranslator(&translator);
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:42]: If hasApp is true, translation will not be applied, will it?

File: src/GLideNUI/GLideNUI.cpp [hunk 8]
```diff
@@ -29,27 +29,34 @@ int openConfigDialog(const wchar_t * _strFileName, const char * _romName, bool &
...
 	w.show();
-	const int res = a.exec();
-#else
-	const int res = w.exec();
-#endif // M64P_GLIDENUI
+
+	int res;
+	if (!hasApp) {
+		res = pApp->exec();
+		delete pApp;
```
Comment (gonetz) [inline src/GLideNUI/GLideNUI.cpp:55]: I also dislike that manual delete pApp.

I suggest something like this:

```
	std::unique_ptr<QApplication> pQApp;
	QCoreApplication* pApp = QCoreApplication::instance();

	if (pApp == nullptr) {
		int argc = 0;
		char * argv = 0;
		pQApp.reset(new QApplication(argc, &argv));
		pApp = pQApp.get();
	}

	QTranslator translator;
	if (translator.load(getTranslationFile(), strIniFileName))
		pApp->installTranslator(&translator);

	ConfigDialog w(Q_NULLPTR, Qt::WindowTitleHint | Qt::WindowSystemMenuHint | Qt::WindowCloseButtonHint);

	w.setIniPath(strIniFileName);
	w.setRomName(_romName);
	w.setTitle();
	w.show();

	int res = pQApp ? pQApp->exec() : w.exec();
	_accepted = w.isAccepted();
	return res;

```

File: src/GLideNUI/GLideNUI.cpp (hunk 6 above)
Comment (Rosalie241) [reply src/GLideNUI/GLideNUI.cpp:54]: well, I'm using `QCoreApplication::instance()` to detect if the GUI (Zilmar spec or mupen spec) is Qt or not, if it is we can't create another `QApplication`

File: src/GLideNUI/GLideNUI.cpp (hunk 7 above)
Comment (Rosalie241) [reply src/GLideNUI/GLideNUI.cpp:42]: that's correct

File: src/GLideNUI/GLideNUI.cpp (hunk 8 above)
Comment (Rosalie241) [reply src/GLideNUI/GLideNUI.cpp]: that works perfectly on both linux & windows, thank you

**Pull request merged**